rust/otap-dataflow/crates/pdata/src/otlp/batching.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use super::OtlpProtoBytes;
+use crate::error::{Error, Result};
+use crate::proto::consts::wire_types;
+use crate::views::otlp::bytes::decode::{read_len_delim, read_varint};
+use otap_df_config::SignalType;
+use std::num::NonZeroU64;
+
+/// Position within the inputs (input index and byte offset)
+#[derive(Copy, Clone, PartialEq)]
+struct Position {
+    input_idx: usize,
+    offset: usize,
+}
+
+impl Position {
+    fn new(input_idx: usize, offset: usize) -> Self {
+        Self { input_idx, offset }
+    }
+}
+
+/// Definition of the batch size in bytes
+fn batch_bytes(start_pos: &Position, end_pos: &Position, inputs: &[OtlpProtoBytes]) -> usize {
+    if start_pos.input_idx == end_pos.input_idx {
+        end_pos.offset - start_pos.offset
+    } else {
+        (inputs[start_pos.input_idx].num_bytes() - start_pos.offset)
+            + (start_pos.input_idx + 1..end_pos.input_idx)
+                .map(|i| inputs[i].num_bytes())
+                .sum::<usize>()
+            + end_pos.offset
+    }
+}
+
+/// Helper to finalize batch by copying bytes from start to end positions
+fn finalize_batch(
+    signal: SignalType,
+    inputs: &[OtlpProtoBytes],
+    start_pos: &Position,
+    end_pos: &Position,
+    outputs: &mut Vec<OtlpProtoBytes>,
+) {
+    // Calculate exact size needed
+    let size = batch_bytes(start_pos, end_pos, inputs);
+
+    if size == 0 {
+        return;
+    }
+
+    let mut batch = Vec::with_capacity(size);
+
+    if start_pos.input_idx == end_pos.input_idx {
+        // Single input
+        let buf = inputs[start_pos.input_idx].as_bytes();
+        batch.extend_from_slice(&buf[start_pos.offset..end_pos.offset]);
+    } else {
+        // Multiple inputs
+        debug_assert!(start_pos.input_idx < inputs.len());
+
+        let buf = inputs[start_pos.input_idx].as_bytes();
+        batch.extend_from_slice(&buf[start_pos.offset..]);
+
+        for req in &inputs[(start_pos.input_idx + 1)..end_pos.input_idx] {
+            batch.extend_from_slice(req.as_bytes());
+        }
+
+        if end_pos.offset != 0 {
+            debug_assert!(end_pos.input_idx < inputs.len());
+            let buf = inputs[end_pos.input_idx].as_bytes();
+            batch.extend_from_slice(&buf[..end_pos.offset]);
+        }
+    }
+
+    outputs.push(OtlpProtoBytes::new_from_bytes(signal, batch))
+}
+
+/// Combines OTLP content by concatenation of bytes. Because we have a
+/// top-level repeated field, this is precisely correct.
```

Comment:
Makes sense. In the future, would we support stepping into the nested Resource<signal> -> Scope<Signal> -> Signal to either to get batch sizes that are more closely aligned to the max size? 

I realize that makes this significantly more complicated, but once we had that logic in place, we could also use it batch OTLP bytes for items sizer (at least for logs & traces).